ethereum/eip712/preprocess.go

```diff
@@ -0,0 +1,79 @@
+package eip712
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	cosmoskr "github.com/cosmos/cosmos-sdk/crypto/keyring"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authtx "github.com/cosmos/cosmos-sdk/x/auth/tx"
+	"github.com/evmos/ethermint/types"
+)
+
+// PreprocessLedgerTx reformats Ledger-signed Cosmos transactions to match the fork expected by Ethermint
+// by including the signature in a Web3Tx extension and sending a blank signature in the body.
+func PreprocessLedgerTx(chainID string, keyType cosmoskr.KeyType, txBuilder client.TxBuilder) error {
+	// Only process Ledger transactions
+	if keyType != cosmoskr.TypeLedger {
+		return nil
+	}
+
+	// Init extension builder to set Web3 extension
+	extensionBuilder, ok := txBuilder.(authtx.ExtensionOptionsTxBuilder)
+	if !ok {
+		return fmt.Errorf("cannot cast TxBuilder to ExtensionOptionsTxBuilder")
+	}
+
+	// Get signatures from TxBuilder
+	sigs, err := txBuilder.GetTx().GetSignaturesV2()
+	if err != nil {
+		return fmt.Errorf("could not get signatures: %w", err)
+	}
+
+	// Verify single-signer
+	if len(sigs) != 1 {
+		return fmt.Errorf("invalid number of signatures, expected 1 and got %v", len(sigs))
+	}
```

Comment:
this might be an issue with native multisigs, have we discussed how to support them?

Reply:
Native multi-sigs shouldn't get to this code, since they should be filtered out by the `TypeLedger` check on line 18. Currently, multi-sig with EIP-712 is not supported node-side, and it would be a separate project to add that capability.